kvdb-web/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A key-value database for use in browsers
+//!
+//! Writes data both into memory and IndexedDB, reads from the IndexedDB
+//! on `open`.
+
+
+mod error;
+mod indexed_db;
+
+use std::io;
+use std::rc::Rc;
+use std::sync::Mutex;
+use kvdb::{DBValue, DBTransaction};
+use kvdb_memorydb::{InMemory, self as in_memory};
+use send_wrapper::SendWrapper;
+
+pub use error::Error;
+pub use kvdb::KeyValueDB;
+
+use futures::prelude::*;
+
+use web_sys::IdbDatabase;
+
+pub struct Database {
+	name: String,
+	version: u32,
+	columns: u32,
+	in_memory: InMemory,
+	indexed_db: Mutex<SendWrapper<IdbDatabase>>,
+}
+
+// The default column is represented as `None`.
+type Column = Option<u32>;
+
+fn number_to_column(col: u32) -> Column {
+	col.checked_sub(1)
+}
+
+
+impl Database {
+	/// Opens the database with the given name,
+	/// and the specified number of columns (not including the default one).
+	pub fn open(name: String, columns: u32) -> impl Future<Output = Result<Database, error::Error>> {
+		// let's try to open the latest version of the db first
+		let open_request = indexed_db::open(name.as_str(), None, columns);
+		let name_clone = name.clone();
+		open_request.then(move |db| {
+			let db = match db {
+				Ok(db) => db,
+				Err(err) => return future::Either::Right(future::err(err)),
+			};
+
+			// if we need more column than the latest version has,
+			// then bump the version (+ 1 for the default column)
+			if columns + 1 > db.columns {
+				let next_version = db.version + 1;
+				drop(db);
+				future::Either::Left(indexed_db::open(name.as_str(), Some(next_version), columns).boxed())
+			} else {
+				future::Either::Left(future::ok(db).boxed())
+			}
+		// populate the in_memory db from the IndexedDB
```

Comment:
This behaviour should be doc'd imo.

Reply:
This is documented in module docs or do you mean something else?

Reply:
Describing it in the module docs is enough.